internal/policies/proxy/proxy.go

```diff
@@ -0,0 +1,138 @@
+// warning and return.
+// If there are entries and ubuntu-proxy-manager is installed, it will call its
+// Apply method via D-Bus service. Any error returned by the Apply call will be
+// returned by the manager.
+//
+// Entry keys passed to the proxy manager that are not part of the supportedKeys
+// list will be ignored and logged as a warning.
+package proxy
+
+import (
+	"context"
+	"os/exec"
+	"strings"
+	"sync"
+
+	"github.com/godbus/dbus/v5"
+	"github.com/ubuntu/adsys/internal/decorate"
+	log "github.com/ubuntu/adsys/internal/grpc/logstreamer"
+	"github.com/ubuntu/adsys/internal/i18n"
+	"github.com/ubuntu/adsys/internal/policies/entry"
+	"github.com/ubuntu/adsys/internal/smbsafe"
+	"golang.org/x/exp/slices"
+)
+
+// Caller is the interface to call a method on a D-Bus object.
+type Caller interface {
+	Call(method string, flags dbus.Flags, args ...interface{}) *dbus.Call
+}
+
+// supportedKeys are the entry keys supported by the proxy manager.
+var supportedKeys = []string{"http", "https", "ftp", "socks", "no-proxy", "auto"}
+
+// Manager prevents running multiple apparmor update processes in parallel while parsing policy in ApplyPolicy.
+type Manager struct {
+	mu           sync.Mutex
+	proxyApplier Caller
+	dpkgCmd      []string
+}
+
+// WithProxyApplier overrides the default proxy applier.
+func WithProxyApplier(c Caller) func(*options) {
+	return func(a *options) {
+		a.proxyApplier = c
+	}
+}
+
+// WithDpkgCmd overrides the default dpkg command.
+func WithDpkgCmd(cmd []string) func(*options) {
+	return func(a *options) {
+		a.dpkgCmd = cmd
+	}
+}
+
+type options struct {
+	proxyApplier Caller
+	dpkgCmd      []string
+}
+
+// Option reprents an optional function to change the proxy manager.
+type Option func(*options)
+
+// New returns a new proxy policy manager.
+func New(bus *dbus.Conn, args ...Option) *Manager {
+	proxyApplier := bus.Object("com.ubuntu.ProxyManager", "/com/ubuntu/ProxyManager")
+
+	// Set default options
+	opts := options{
+		proxyApplier: proxyApplier,
+		dpkgCmd:      []string{"dpkg", "--status"},
+	}
+
+	// Apply given options
+	for _, f := range args {
+		f(&opts)
+	}
+
+	return &Manager{
+		proxyApplier: opts.proxyApplier,
+		dpkgCmd:      opts.dpkgCmd,
+	}
+}
+
+// ApplyPolicy applies the system proxy policy (via a D-Bus call to ubuntu-proxy-manager).
+func (m *Manager) ApplyPolicy(ctx context.Context, objectName string, isComputer bool, entries []entry.Entry) (err error) {
+	defer decorate.OnError(&err, i18n.G("can't apply proxy policy"))
+
+	// Proxy policies are currently only supported on computers
+	if !isComputer {
+		return nil
+	}
+
+	// Exit early if we don't have any entries to apply
+	if len(entries) == 0 {
+		return nil
+	}
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	dpkgStatusCmd := append(m.dpkgCmd, "ubuntu-proxy-manager")
+
+	smbsafe.WaitExec()
+	// #nosec G204 - We are in control of the arguments
+	out, err := exec.CommandContext(ctx, dpkgStatusCmd[0], dpkgStatusCmd[1:]...).CombinedOutput()
```

Comment:
This is nice, but I wouldn’t add really dpkg knowledge here to the code base. I would just execute the dbus call (as if it was any other binaries) and returns an errors.

Or, if you want to be nicer as it’s a suggest, you can try to see if the hardcoded path is there, but really… unsure. I think the documentation can cover it.

It’s a little bit similar than systemd (even if I doubt we will run on Devuan anytime soon), or without dconf, for servers…